src/Juvix/Frontend/Types/Base.hs

```diff
@@ -0,0 +1,443 @@
+
+type NameSymb = NonEmpty Symbol
+
+type ModuleName = NameSymb
+
+extensible
+  [d|
+    data TopLevel
+      = Type Type
+      | ModuleOpen ModuleOpen
+      | Signature Signature
+      | Module Module
+      | Function Function
+      | TypeClass
+      | TypeClassInstance
+      deriving (Show, Generic, NFData)
+
+    --------------------------------------------------------------------------------
+    -- Types
+    --------------------------------------------------------------------------------
+
+    data Type
+      = Typ
+          -- Was a usage but can't alias for now
+          { typeUsage :: Maybe Expression,
+            typeName' :: !Symbol,
+            typeArgs :: [Symbol],
+            typeForm :: TypeSum
+          }
+      deriving (Show, Generic, NFData)
+
+    data TypeSum
+      = Alias Alias
+      | Data Data
+      deriving (Show, Generic, NFData)
+
+    -- | 'Data' is the data declaration in the Juvix language
+    data Data
+      = Arrowed
+          { dataArrow :: Expression,
+            dataAdt' :: Adt
+          }
+      | NonArrowed
+          { dataAdt :: Adt
+          }
+      deriving (Show, Generic, NFData)
+
+    -- was newtype
+    data Alias
+      = AliasDec
+          {aliasType' :: Expression}
+      deriving (Show, Generic, NFData)
+
+    --------------------------------------------------
+    -- Arrows
+    --------------------------------------------------
+    data NamedType
+      = NamedType'
+          { nameRefineName :: !Name,
+            namedRefineRefine :: Expression
+          }
+      deriving (Show, Generic, NFData)
+
+    -- TODO ∷ change TypeName to TypeNameModule
+    data TypeRefine
+      = TypeRefine
+          { typeRefineName :: Expression,
+            typeRefineRefinement :: Expression
+          }
+      deriving (Show, Generic, NFData)
+
+    --------------------------------------------------
+    -- Types Misc
+    --------------------------------------------------
+
+    data Name
+      = Implicit !Symbol
+      | Concrete !Symbol
+      deriving (Show, Generic, NFData)
+
+    data ArrowSymbol
+      = ArrowUse Usage.T
+      | -- Was a usage but can't alias for now
+        ArrowExp Expression
+      deriving (Show, Generic, NFData)
+
+    -- Was a new type
+    -- TODO ∷ finish this type!
+    data UniverseExpression
+      = UniverseExpression Symbol
+      deriving (Show, Generic, NFData)
+
+    --------------------------------------------------
+    -- ADTs
+    --------------------------------------------------
+
+    data Adt
+      = Sum (NonEmpty Sum)
+      | Product Product
+      deriving (Show, Generic, NFData)
+
+    data Sum
+      = S
+          { sumConstructor :: !Symbol,
+            sumValue :: !(Maybe Product)
+          }
+      deriving (Show, Generic, NFData)
+
+    data Product
+      = Record !Record
+      | Arrow Expression
+      | ADTLike [Expression]
+      deriving (Show, Generic, NFData)
+
+    data Record
+      = Record''
```

Comment:
I wonder if we should maybe come up with a different naming scheme for these things than variously having a single/double `'`?

Maybe like `RecordInfo` or something, but I'm not that fussed about what exactly if someone has a better suggestion.

Reply:
Yes  I would like to do that at some point